pipeline/07-export.R

```diff
@@ -285,6 +283,74 @@ assessment_pin_merged <- assessment_pin %>%
   filter(!is.na(pred_pin_final_fmv_land)) %>%
   mutate(across(ends_with("_date"), as_date))
 
+if (comp_enable) {
+  message("Pulling comp data from Athena")
+  # We don't know how many comp_score_{x} columns there are, and selecting
+  # columns via wildcard in SQL is complicated, so just query everything and
+  # then select the necessary columns in R
+  comps <- dbGetQuery(
+    conn = AWS_ATHENA_CONN_NOCTUA, glue("
+    SELECT *
+    FROM model.comp
+    WHERE run_id = '{params$export$run_id}'
+    ")
+  ) %>%
+    mutate(
+      overall_comp_score = select(., starts_with("comp_score_")) %>%
+        rowMeans(na.rm = TRUE)
+    ) %>%
+    select(
+      pin, overall_comp_score,
+      comp_pin_1, comp_score_1, comp_pin_2, comp_score_2
+    )
+
+  # Merge comp data into assessment data. Start with single-card PINs, where
+  # the comps for the card are the same as the comps for the parcel
+  assessment_pin_single_card <- assessment_pin_merged_w_land %>%
+    filter(meta_pin_num_cards == 1 | is.na(meta_pin_num_cards)) %>%
+    left_join(comps, by = join_by(meta_pin == pin))
+
+  # In cases where the PIN has multiple cards, choose the highest-performing
+  # comps across all cards as the top two. This is necessary because comps are
+  # defined at the card level and not the parcel level.
+  # To start, group the comps by PIN and select the comps with the top
+  # overall_score. This is ugly datatable code but it is much faster than the
+  # dplyr version using group_by and slice_max
+  comps_by_top_score <- comps[
+    comps[, .I[overall_comp_score == max(overall_comp_score)], by = pin]$V1
+  ] %>%
+    distinct()
+  # Next, join the PIN-level data to the top-scoring comps
+  assessment_pin_multicard <- assessment_pin_merged_w_land %>%
+    filter(meta_pin_num_cards > 1) %>%
+    left_join(comps_by_top_score, by = join_by(meta_pin == pin))
```

Comment:
Note that this isn't exactly the behavior we discussed; notably, this code takes the top two comps that correspond to the highest overall score, and uses those for the target PIN.

The reason I went in this direction is that I implemented the version with the double-pivot that gets the top-2 comps among all cards on the PIN, before I realized that it has three key flaws:

1. The dplyr code is extremely slow compared to the dataframe code here
2. The overall score does not necessarily match to the comps that are selected
3. If the same PIN is a high-performing comparable for both cards, it can be selected twice

I found myself going down a number of rabbit holes trying to fix all three of these problems, so I paused and decided to go with the simpler solution instead to unblock myself. If you can come up with some dplyr code that fixes all of these problems, I would love to drop it in!

Reply:
I _think_ this tracks but we may have to do a further look at this down the line. Let's go with it for now though.